### PR TITLE
feat(core): introduce `afterRenderEffect`

### PR DESCRIPTION
**Pull request by alxhub**
Implement the `afterRenderEffect` primitive, which creates effect(s) that run as part of Angular's `afterRender` sequence. `afterRenderEffect` is a useful primitive for expressing DOM operations in a declarative, reactive way.

The API itself mirrors `afterRender` and `afterNextRender` with one big difference: values are propagated from phase to phase as signals instead of as plain values. As a result, later phases may not need to execute if the values returned by earlier phases do not change.

### CODE REVIEW

**Comment by garrettld** [comment]
I love this change, but since the signalification of the primitive values returned by each phase callback is abstracted away from us, this seems to be a bad fit for use cases that require multiple values to be passed from one phase to another. For example:

```ts
afterRenderEffect({
  earlyRead: () => {
    const element = this.someElementRef.nativeElement;
    return [element.offsetLeft, element.offsetTop];
  },
  write: ([x, y]) => animateThing(x, y)
})
```

In this example, I assume that the write phase would run on every render because the earlyRead phase returns a new array each time. Is that correct?

The workaround I use today isn't very ergonomic:

```ts
position = signal<[number,number]>([0,0], {equal: arrayDeepEqual});

afterRender({
  earlyRead: () => {
    const element = this.someElementRef().nativeElement;
    this.position.set([element.offsetLeft, element.offsetTop]);
  }
});

// only run the write phase when this.position changes, after accounting for custom equality fn
effect(() => {
  const [x, y] = this.position();
  afterNextRender({ write: () => animateThing(x, y) });
});
```

It seems to work, but it feels brittle. What do you think?

File: packages/core/src/render3/reactivity/after_render_effect.ts [hunk 1]
```diff
@@ -0,0 +1,381 @@
...
+    }
+
+    // Run any needed cleanup functions.
+    try {
+      for (const cleanupFn of this.cleanup ?? EMPTY_CLEANUP_SET) {
+        cleanupFn();
+      }
+    } finally {
+      // Even if a cleanup function errors, ensure it's cleared.
+      this.cleanup?.clear();
```
Comment (atscott) [inline packages/core/src/render3/reactivity/after_render_effect.ts:116]: Does this mean that if one cleanup function errors, the others don't run but are still cleared? 

Should this be more like 


```
for ... {
  try {
    cleanupFn();
  } finally {
    this.cleanup?.delete(cleanupFn);
 }
}
```

File: packages/core/test/acceptance/after_render_effect_spec.ts [hunk 2]
```diff
@@ -0,0 +1,336 @@
...
+  ApplicationRef,
+  computed,
+  PLATFORM_ID,
+  provideExperimentalZonelessChangeDetection,
+  signal,
+} from '@angular/core';
+import {afterRenderEffect} from '@angular/core/src/render3/reactivity/after_render_effect';
+import {TestBed} from '@angular/core/testing';
+
+describe('afterRenderEffect', () => {
```
Comment (pkozlowski-opensource) [inline packages/core/test/acceptance/after_render_effect_spec.ts:20]: Very comprehensive and easy to follow tests - great job here 👍

File: packages/core/test/acceptance/after_render_effect_spec.ts [hunk 3]
```diff
@@ -0,0 +1,336 @@
...
+    );
+
+    appRef.tick();
+    log.length = 0;
+
+    counter.set(1);
+    appRef.tick();
+    expect(log).toEqual(['earlyRead: 1', 'earlyRead: 0', 'write: 0']);
+  });
+});
```
Comment (pkozlowski-opensource) [inline packages/core/test/acceptance/after_render_effect_spec.ts:336]: I'm not sure I've seen a test where one phase (ex.: early read) runs and returns a value but it is the same as the previous one - in the case a later phase depending only on this previous value shouldn't run, if I'm not mistaken.

**Review by pkozlowski-opensource — APPROVED** [review]

**Comment by pkozlowski-opensource** [comment]
LGTM overall but adding a cleanup label since:
* is missing public API export
* it has pending comments from other reviewers

File: packages/core/test/acceptance/after_render_effect_spec.ts (hunk 3 above)
Comment (alxhub) [reply packages/core/test/acceptance/after_render_effect_spec.ts:336]: "should pass data from one phase to the next via signal" tests both sides of that proposition - that propagation works, and that the later phase doesn't rerun if the data doesn't change.

File: packages/core/src/render3/reactivity/after_render_effect.ts (hunk 1 above)
Comment (alxhub) [reply packages/core/src/render3/reactivity/after_render_effect.ts:116]: I thought about this before, and chose (for no strong reason other than intuition) to implement a mental model where we considered all the registered cleanup functions to be a part of the same logical operation. We really need to standardize this model (`effect()` for example overwrites the previous cleanup function when a new one is registered).

**Comment by alxhub** [comment]
@garrettld
> I love this change, but since the signalification of the primitive values returned by each phase callback is abstracted away from us, this seems to be a bad fit for use cases that require multiple values to be passed from one phase to another.

Yeah, we saw this too, but I don't know if there is an API that's ideal here. There seem to be two potential design avenues which address the problem:

1. Allowing some configuration of equality function (which users would have to remember to do). This could be specified in the options, or by allowing the signal to be directly created by the user somehow.

2. Increasing the granularity at which we make signals out of the returned object. For example, we could require that phases return an object & make signals out of each property, or explicitly support returning tuples.

Both of these solutions have their downsides. With equality functions, you have to remember to pass them. With more granular return objects, the signal wrapping only works at one layer - if I get creative and `return {bounds: {x: 0, y: 3}}` then I break things again. I think we're leaning in favor of equality as a more explicit solution.

**Comment by AndrewKushnir** [comment]
@alxhub the `test` CI job is failing with the payload size issue. Could you please update golden file(s)?

**Review by thePunderWoman — APPROVED** [review]
reviewed-for: public-api

**Review by AndrewKushnir — APPROVED** [review]
Reviewed-for: public-api

**Review by AndrewKushnir — APPROVED** [review]
Reviewed-for: size-tracking

**Comment by AndrewKushnir** [comment]
This PR was merged into the repository by commit be2e49639bda831831ad62d49253db942a83fd46.

The changes were merged into the following branches: main

**Pull request closed without merging**